File: 413-arithmetic-slices.cs
public class Solution {
    public int NumberOfArithmeticSlices(int[] A) {
        var dp = new int[A.Length];
        var sum = 0;
        for (var i = 2; i < dp.Length; i++)
            if (A[i] - A[i - 1] == A[i - 1] - A[i - 2]) {
                dp[i] = 1 + dp[i - 1];
                sum += dp[i];
            }

        return sum;
    }
}

File: 413-arithmetic-slices.go
package main

func numberOfArithmeticSlices(nums []int) int {
	n := len(nums)
	res := 0
	for i := 0; i < n-2; {
		j := i
		diff := nums[i+1] - nums[i]
		for ; j+1 < n && nums[j+1]-nums[j] == diff; j++ {
		}
		length := j - i + 1
		a1 := 1
		an := length - 2
		res += (a1 + an) * an / 2
		i = j
	}
	return res
}
